### PR TITLE
[FEATURE] Adds dynamic curly components and helpers

### PR DESCRIPTION
**Pull request by pzuraq**
Adds dynamic curly component and helper invocation, as well as the
ability to curry components with the `(helper)` keyword.

### CODE REVIEW

File: packages/@glimmer/compiler/lib/passes/1-normalization/keywords/append.ts [hunk 1]
```diff
@@ -251,4 +251,62 @@ export const APPEND_KEYWORDS = keywords('Append')
...
+      if (definition === null) {
+        return Err(
+          generateSyntaxError(
+            `{{helper}} requires a helper definition as its first positional parameter, did not receive any parameters.`,
+            args.loc
+          )
+        );
+      }
+
+      if (state.isStrict && definition.type === 'Literal') {
```
Comment (rwjblue) [inline packages/@glimmer/compiler/lib/passes/1-normalization/keywords/append.ts:275]: As discussed, should be

```suggestion
      if (definition.type === 'Literal') {
```

File: packages/@glimmer/integration-tests/test/strict-mode-test.ts [hunk 2]
```diff
@@ -646,6 +737,19 @@ class DynamicStrictModeTest extends RenderTest {
...
+    const Bar = defineComponent(
+      {},
+      '<div {{this.foo 123}}></div>',
+      class extends GlimmerishComponent {}
+    );
+
+    this.renderComponent(Bar);
+    this.assertHTML('');
+    this.assertStableRerender();
+  }
```
Comment (rwjblue) [inline packages/@glimmer/integration-tests/test/strict-mode-test.ts:751]: This is just future work, right?

File: packages/@glimmer/vm/lib/opcodes.toml [hunk 3]
```diff
@@ -187,6 +196,18 @@ operand-stack = [
   ["unknown"]
 ]
 
+[syscall.rconstrefload]
+
+format = ["ConstantReference", "constant:unknown"]
+operation = """
+  Push an reference constant onto the stack that is not
```
Comment (rwjblue) [inline packages/@glimmer/vm/lib/opcodes.toml:203]:
```suggestion
  Push a reference constant onto the stack that is not
```

File: packages/@glimmer/integration-tests/test/strict-mode-test.ts (hunk 2 above)
Comment (pzuraq) [reply packages/@glimmer/integration-tests/test/strict-mode-test.ts:751]: Yes, that's why the test was skipped, was just easier to write it out now while I had the context

File: packages/@glimmer/syntax/lib/v2-a/loose-resolution.ts [hunk 4]
```diff
@@ -76,7 +76,7 @@ export function AppendSyntaxContext(node: ASTv1.MustacheStatement): ASTv2.FreeVa
       ? ASTv2.LooseModeResolution.trustingAppend({ invoke: isInvoke })
       : ASTv2.LooseModeResolution.append({ invoke: isInvoke });
   } else {
-    return isInvoke ? ASTv2.STRICT_RESOLUTION : ASTv2.LooseModeResolution.fallback();
+    return ASTv2.LooseModeResolution.fallback();
```
Comment (rwjblue) [inline packages/@glimmer/syntax/lib/v2-a/loose-resolution.ts:79]: Hmm, does this introduce property fallback behavior?

**Review by rwjblue — COMMENTED** [review]
Looks good, thanks for fixing issues with helper invocation from local paths!

I think we need some tests for those cases though (previously it was impossible).

File: packages/@glimmer/syntax/lib/v2-a/loose-resolution.ts (hunk 4 above)
Comment (rwjblue) [reply packages/@glimmer/syntax/lib/v2-a/loose-resolution.ts:79]: Chatted with @pzuraq, we already have tests confirming implicit this fallback does **not** happen for strict.

@pzuraq - if you have a chance, would you mind linking to that test?

**Comment by rwjblue** [comment]
Chatted with @pzuraq and I think we need to add a couple more test cases for the new loose mode capability introduced here:

```hbs
{{foo.bar "derp"}}
```

* loose mode, `foo` is yielded block param with a `bar` prop that is a helper
* loose mode, `{{foo.bar "derp"}}` resolves to `this.foo.bar`  (and can be invoked)

For strict mode, we already have these tests:

* strict mode, foo is a local in scope that has a bar prop that is a helper:
* strict mode, foo is not a local in scope (should error)

File: packages/@glimmer/syntax/lib/v2-a/loose-resolution.ts (hunk 4 above)
Comment (pzuraq) [reply packages/@glimmer/syntax/lib/v2-a/loose-resolution.ts:79]: In Ember, this previously would have been compiled to `{{component foo.bar baz=123}}` or something like that, so I think the ship has sailed in that regard. That's why I included the change here.

**Review by rwjblue — APPROVED** [review]

**Pull request merged**